cmd/health_v1.go

```diff
@@ -0,0 +1,272 @@
+	Partitions []madmin.PartitionStat         `json:"partitions,omitempty"`
+	Usage      []*disk.UsageStat              `json:"usage,omitempty"`
+	Error      string                         `json:"error,omitempty"`
+}
+
+// HwMemV1 - Includes host virtual and swap mem information
+type HwMemV1 struct {
+	SwapMem    *mem.SwapMemoryStat    `json:"swap,omitempty"`
+	VirtualMem *mem.VirtualMemoryStat `json:"virtualmem,omitempty"`
+	Error      string                 `json:"error,omitempty"`
+}
+
+// HwPerfV1 - hardware performance
+type HwPerfV1 struct {
+	Net   HwNetPerfV1   `json:"net,omitempty"`
+	Drive HwDrivePerfV1 `json:"drives,omitempty"`
+}
+
+// HwNetPerfV1 - Network performance info
+type HwNetPerfV1 struct {
+	Serial   []madmin.NetPerfInfo `json:"serial,omitempty"`
+	Parallel []madmin.NetPerfInfo `json:"parallel,omitempty"`
+}
+
+// HwDrivePerfV1 - Network performance info
+type HwDrivePerfV1 struct {
+	Serial   []madmin.DrivePerfInfo `json:"serial,omitempty"`
+	Parallel []madmin.DrivePerfInfo `json:"parallel,omitempty"`
+	Error    string                 `json:"error,omitempty"`
+}
+
+// SwInfoV1 - software health Info
+type SwInfoV1 struct {
+	Minio  MinioHealthInfoV1     `json:"minio,omitempty"`
+	OsInfo []madmin.ServerOsInfo `json:"osinfos,omitempty"`
+}
+
+// MinioHealthInfoV1 - Health info of the MinIO cluster
+type MinioHealthInfoV1 struct {
+	Info     madmin.InfoMessage      `json:"info,omitempty"`
+	Config   interface{}             `json:"config,omitempty"`
+	ProcInfo []madmin.ServerProcInfo `json:"procinfos,omitempty"`
+	Error    string                  `json:"error,omitempty"`
+}
+
+// ClusterHealthV1 - main struct of the health report
+type ClusterHealthV1 struct {
+	TimeStamp time.Time   `json:"timestamp,omitempty"`
+	Status    string      `json:"status"`
+	Error     string      `json:"error,omitempty"`
+	Hardware  HwServersV1 `json:"hardware,omitempty"`
+	Software  SwInfoV1    `json:"software,omitempty"`
+}
+
+func (ch ClusterHealthV1) String() string {
+	return ch.JSON()
+}
+
+// JSON jsonifies service status message.
+func (ch ClusterHealthV1) JSON() string {
+	statusJSONBytes, e := json.MarshalIndent(ch, " ", "    ")
+	fatalIf(probe.NewError(e), "Unable to marshal into JSON.")
+
+	return string(statusJSONBytes)
+}
+
+// GetStatus - return status of the health info
+func (ch ClusterHealthV1) GetStatus() string {
+	return ch.Status
+}
+
+// GetError - return error from the health info
+func (ch ClusterHealthV1) GetError() string {
+	return ch.Error
+}
+
+// GetTimestamp - return timestamp from the health info
+func (ch ClusterHealthV1) GetTimestamp() time.Time {
+	return ch.TimeStamp
+}
+
+func mapHealthInfoToV1(healthInfo madmin.HealthInfo, err error) HealthReportInfo {
+	ch := ClusterHealthV1{}
+	ch.TimeStamp = healthInfo.TimeStamp
+	if err != nil {
+		ch.Status = "Error"
+		ch.Error = err.Error()
+		return ch
+	}
+
+	ch.Status = "Success"
+
+	serverAddrs := set.NewStringSet()
+
+	serverCPUs := mapServerCPUs(healthInfo)
```

Comment:
Suggestion (Ignore if you don't think it makes much difference here) - Can we parallelize the map calls here..?

Reply:
Something like the following should work If we decide to parallelize. 

```
var serverCPUs map[string][]HwCPUV1
var serverDrives map[string][]HwDriveV1
var serverMems map[string]HwMemV1
..
..
fetchServerCPUFunc := func() {
    serverCPUs = mapServerCPUs(healthInfo)
}
fetchServerDrivesFunc = func() {
    serverDrives = mapServerDrives(healthInfo)
}
..
..
Parallelize(fetchServerCPUFunc, fetchServerDrivesFunc, ..) // Will execute the functions and wait
..
..
// Parallelize parallelizes the function calls
func Parallelize(functions ...func()) {
    var waitGroup sync.WaitGroup
    waitGroup.Add(len(functions))

    defer waitGroup.Wait()

    for _, function := range functions {
        go func(fn func()) {
            defer waitGroup.Done()
             fn()
        }(function)
    }
}
```